### PR TITLE
feat(icon): add general kirby icons and documentation how to update them

### LINKED ISSUE #315: Add all general icons
**Is your enhancement request related to a problem? Please describe.**
We would like to add all general icons, like navigation, link, settings etc. so that everyone can use them in their respective areas. So that we use the same icons cross platform.

See also: https://app.zeplin.io/project/5c99d69a79d02b35ea3b8f41/styleguide/components?cseid=5ca73870a994f9598bccef93

### PR DESCRIPTION
**Pull request by hafstad**
closes #315
- General icons has been added
- Icon component / examples has been refactored
- Custom-icon directive has been baked into the icon-component as it was the simplest

### CODE REVIEW

File: src/kirby/components/icon/custom-icon-settings.ts [hunk 1]
```diff
@@ -1,8 +1,15 @@
 import { InjectionToken } from '@angular/core';
 
+export interface CustomIcon {
```
Comment (jakobe) [inline src/kirby/components/icon/custom-icon-settings.ts]: Guess it's just an Icon definition now (not really custom anymore, as it's used by Kirby itself)
```suggestion
export interface Icon {
```

File: src/kirby/components/icon/custom-icon-settings.ts [hunk 2]
```diff
@@ -1,8 +1,15 @@
...
+  name: string;
+  svg?: string;
+  unicode: string;
+  fromCharCode?: string;
```
Comment (jakobe) [inline src/kirby/components/icon/custom-icon-settings.ts]: Consider naming the property as is (the `fromCharCode` is the method used to generate the prop):
```suggestion
  charCode?: string;
```

File: src/kirby/components/icon/custom-icon-settings.ts [hunk 3]
```diff
@@ -1,8 +1,15 @@
...
+}
+
 export interface CustomIconSettings {
```
Comment (jakobe) [inline src/kirby/components/icon/custom-icon-settings.ts]: Consider renaming:
```suggestion
export interface IconSettings {
```

File: src/kirby/components/icon/custom-icon-settings.ts [hunk 4]
```diff
@@ -1,8 +1,15 @@
...
   fontfamily: string;
-  icons: Array<{ name: string; svg: string; unicode: string }>;
+  icons: CustomIcon[];
 }
 
 export const CUSTOM_FONT_SETTINGS = new InjectionToken<CustomIconSettings>('CustomIconSettings');
```
Comment (jakobe) [inline src/kirby/components/icon/custom-icon-settings.ts]: Consider renaming:
```suggestion
export const ICON_SETTINGS = new InjectionToken<IconSettings>('IconSettings');
```

File: src/kirby/components/icon/kirby-custom-icon-settings.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import { CustomIconSettings } from '@kirbydesign/designsystem/components/icon/custom-icon-settings';
+
+// tslint:disable:prettier
+export const kirbyCustomIconSettings: CustomIconSettings = {
```
Comment (jakobe) [inline src/kirby/components/icon/kirby-custom-icon-settings.ts]: Consider renaming:
```suggestion
export const iconSettings: IconSettings = {
```

File: src/kirby/components/icon/icon.component.ts [hunk 6]
```diff
@@ -10,55 +13,67 @@ import { Component, Input, HostBinding } from '@angular/core';
...
+      }
+    }
 
-  getIonicIconName(name: string): string {
-    const icon = iconsCharCodeMap[name];
-    return icon !== undefined ? name : IconComponent.DEFAULT_ICON_NAME;
+    if (icon) {
+      this._icon = {
+        ...icon,
+        fromCharCode: this.fromCharCode(icon.unicode),
```
Comment (jakobe) [inline src/kirby/components/icon/icon.component.ts]: Consider inlining (1 line method wrapper):
```suggestion
        fromCharCode: String.fromCharCode(icon.unicode),
```

File: src/kirby/components/icon/icon.component.ts [hunk 7]
```diff
@@ -10,55 +13,67 @@ import { Component, Input, HostBinding } from '@angular/core';
...
+        changes.customName.currentValue
+      );
+    }
+  }
+
+  private fromCharCode(icon) {
+    return String.fromCharCode(icon);
+  }
+
+  private getCustomIcon(icons, name: string): CustomIcon {
```
Comment (jakobe) [inline src/kirby/components/icon/icon.component.ts]: Add type:
```suggestion
  private getCustomIcon(icons: CustomIcon[], name: string): CustomIcon {
```

Comment (jakobe) [inline src/kirby/components/icon/icon.component.ts]: Consider renaming:
```suggestion
  private getIcon(icons, name: string): CustomIcon {
```

File: src/kirby/components/icon/icon.component.ts [hunk 8]
```diff
@@ -10,55 +13,67 @@ import { Component, Input, HostBinding } from '@angular/core';
...
+  private _icon = (this.icon = this.defaultIcon);
+
+  @Input() name: string;
+  @Input() customName: string;
+
+  get icon(): CustomIcon {
+    return this._icon;
+  }
+
+  set icon(icon: CustomIcon) {
```
Comment (jakobe) [inline src/kirby/components/icon/icon.component.ts]: A bit hard to follow this code block...

File: src/app/examples/examples.common.ts [hunk 9]
```diff
@@ -23,7 +23,7 @@ import { SegmentedChipControlExampleComponent } from './segmented-chip-control-e
 import { ActionSheetExampleComponent } from './action-sheet-example/action-sheet-example.component';
 import { AlertExampleComponent } from './alert-example/alert-example.component';
 
-export const customIconSettings: CustomIconSettings = {
+export const iconSettings: IconSettings = {
   fontfamily: "'Ionicons', 'ionicons'",
```
Comment (jakobe) [inline src/app/examples/examples.common.ts]: :fire: Remove Ionicons

File: src/app/examples/examples.common.ts [hunk 10]
```diff
@@ -64,6 +64,4 @@ export const COMPONENT_DECLARATIONS: any[] = [
   AlertExampleComponent,
 ];
 
-export const PROVIDER_DECLARATIONS: any[] = [
-  { provide: CUSTOM_FONT_SETTINGS, useValue: customIconSettings },
-];
+export const PROVIDER_DECLARATIONS: any[] = [{ provide: ICON_SETTINGS, useValue: iconSettings }];
```
Comment (jakobe) [inline src/app/examples/examples.common.ts]: Either remove (Kirby icons should come out of the box) or replace with example on custom icon settings/font

File: src/app/examples/icon-example/icon-example.component.html [hunk 11]
```diff
@@ -1,36 +1,44 @@
...
-<kirby-icon name="search" title="search"></kirby-icon>
-<kirby-icon name="checkbox" title="checkbox"></kirby-icon>
-<kirby-icon name="checkbox-outline" title="checkbox-outline"></kirby-icon>
-<kirby-icon name="menu" title="menu"></kirby-icon>
-<kirby-icon name="person" title="person"></kirby-icon>
+<p class="note">*If icon is <strong>NOT</strong> found, 'cog' will be used as default icon</p>
+
+<div class="icon-item-container" *ngFor="let icon of icons">
+    <div class="icon-item-inner-container">
+        <kirby-icon [name]="icon.name" size="md" title="md"></kirby-icon>
```
Comment (jakobe) [inline src/app/examples/icon-example/icon-example.component.html]: Remove size and use icon.name as title:
```suggestion
        <kirby-icon [name]="icon.name" [title]="icon.name"></kirby-icon>
```

File: src/kirby/index.ts [hunk 12]
```diff
@@ -5,15 +5,14 @@ export { CardFooterComponent } from './components/card/card-footer/card-footer.c
...
-export { CustomIconNameDirective } from './components/icon/custom-icon-name.directive';
 export { KirbyModule } from './kirby.module';
 export { SpinnerComponent } from './components/spinner/spinner.component';
 export {
   SegmentedControlComponent,
 } from './components/segmented-control/segmented-control.component';
 export { SegmentItem } from './components/segmented-control/segment-item';
 export { BadgeComponent } from './components/badge/badge.component';
-export { CustomIconSettings, CUSTOM_FONT_SETTINGS } from './components/icon/custom-icon-settings';
+export { IconSettings, ICON_SETTINGS } from './components/icon/icon-settings';
```
Comment (jakobe) [inline src/kirby/index.ts]: Consider moving together with/after `export { IconComponent } from './components/icon/icon.component';`

File: src/kirby/components/icon/icon.component.scss [hunk 13]
```diff
@@ -5,6 +5,11 @@
...
   height: 1em;
   font-size: size('m');
+
+  svg path {
+    fill: currentColor;
+    stroke: currentColor;
+  }
 }
 
 :host(.sm) {
```
Comment (bdyken) [inline src/kirby/components/icon/icon.component.scss]: how about when using icons in a badge ? should we not have a size for badge, the example on master the icon do not fit the badge on native and we might also need a smaller icon for "normal" size badge, so we have the size that are shown on the cookbook and one for the size of the texts examples, is that considered here ?

File: src/kirby/icons/svg/QR.svg [hunk 14]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<svg width="24px" height="24px" viewBox="0 0 24 24" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+    <!-- Generator: Sketch 55.1 (78136) - https://sketchapp.com -->
+    <title>QR</title>
```
Comment (bdyken) [inline src/kirby/icons/svg/QR.svg]: there is a slide problem with title and description, when using the icon om web. The title for the svg is show when mouse over is active, this override the mouse over for any parent element that are set. We have solved this with remove the title and decs for the svg file

File: src/kirby/components/icon/icon.component.scss (hunk 13 above)
Comment (jakobe) [reply src/kirby/components/icon/icon.component.scss]: @bdyken It's generally considered bad practice to comment on code/features outside the PR that the requester is asking for a review on. You should create  a GH issue (bug or feature request) instead. A PR/review is about the changes made, not the overall component/feature in question.
